File: helfomat-api/src/test/java/de/helfenkannjeder/helfomat/api/organization/TravelDistanceApplicationServiceTest.java
package de.helfenkannjeder.helfomat.api.organization;

import de.helfenkannjeder.helfomat.api.geopoint.DistanceMatrixApplicationService;
import de.helfenkannjeder.helfomat.api.geopoint.TravelDistanceDto;
import de.helfenkannjeder.helfomat.api.geopoint.TravelModeDto;
import de.helfenkannjeder.helfomat.core.geopoint.GeoPoint;
import de.helfenkannjeder.helfomat.core.organization.Address;
import de.helfenkannjeder.helfomat.core.organization.Organization;
import de.helfenkannjeder.helfomat.core.organization.OrganizationId;
import de.helfenkannjeder.helfomat.core.organization.OrganizationRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.AdditionalMatchers.not;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class TravelDistanceApplicationServiceTest {

    @Mock
    private DistanceMatrixApplicationService distanceMatrixService;

    @Mock
    private OrganizationRepository organizationRepository;

    private TravelDistanceApplicationService travelDistanceApplicationService;

    private static final TravelDistanceDto DUMMY_TRAVEL_DISTANCE = new TravelDistanceDto();

    @BeforeEach
    void setUp() {
        travelDistanceApplicationService = new TravelDistanceApplicationService(distanceMatrixService, organizationRepository);
    }

    static {
        DUMMY_TRAVEL_DISTANCE.setTravelMode(TravelModeDto.CYCLING);
        DUMMY_TRAVEL_DISTANCE.setDistanceInMeters(34200L);
        DUMMY_TRAVEL_DISTANCE.setTimeInSeconds(31200L);
    }

    @Test
    void searchReturnsValuesForAllPossibleTravelOptions() {
        when(distanceMatrixService.getTravelDistanceFor(any(), any(), any())).thenReturn(DUMMY_TRAVEL_DISTANCE);

        List<TravelDistanceDto> travelOptions = testSearch();

        assertThat(travelOptions)
            .isNotNull()
            .hasSize(TravelModeDto.values().length);
    }

    @Test
    void onlyReturnDistancesForFoundRoutes() {
        when(distanceMatrixService.getTravelDistanceFor(eq(TravelModeDto.TRANSIT), any(), any())).thenReturn(null);
        when(distanceMatrixService.getTravelDistanceFor(not(eq(TravelModeDto.TRANSIT)), any(), any())).thenReturn(DUMMY_TRAVEL_DISTANCE);

        List<TravelDistanceDto> travelOptions = testSearch();

        assertThat(travelOptions)
            .isNotNull()
            .hasSize(TravelModeDto.values().length - 1);
        assertThat(
            travelOptions.stream()
                .filter(travelOption -> travelOption.getTravelMode() == TravelModeDto.TRANSIT)
                .collect(Collectors.toList())
        )
            .isEmpty();
    }

    private List<TravelDistanceDto> testSearch() {
        Address address = new Address.Builder()
            .setLocation(new GeoPoint(49.0388109, 8.3433651))
            .build();
        OrganizationId organizationId = new OrganizationId();
        when(organizationRepository.findOne(organizationId.getValue())).thenReturn(new Organization.Builder()
            .setDefaultAddress(address)
            .build());

        return travelDistanceApplicationService.requestTravelDistances(organizationId, new GeoPoint(48.9808278, 8.4907565));
    }

}

File: helfomat-api/src/test/kotlin/de/helfenkannjeder/helfomat/api/organization/TravelDistanceApplicationServiceTest.kt
package de.helfenkannjeder.helfomat.api.organization

import de.helfenkannjeder.helfomat.api.geopoint.DistanceMatrixApplicationService
import de.helfenkannjeder.helfomat.api.geopoint.TravelDistanceDto
import de.helfenkannjeder.helfomat.api.geopoint.TravelModeDto
import de.helfenkannjeder.helfomat.core.geopoint.GeoPoint
import de.helfenkannjeder.helfomat.core.organization.Address
import de.helfenkannjeder.helfomat.core.organization.Organization
import de.helfenkannjeder.helfomat.core.organization.OrganizationId
import de.helfenkannjeder.helfomat.core.organization.OrganizationRepository
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.AdditionalMatchers.not
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
internal class TravelDistanceApplicationServiceTest {
    @Mock
    private lateinit var distanceMatrixService: DistanceMatrixApplicationService

    @Mock
    private lateinit var organizationRepository: OrganizationRepository
    private lateinit var travelDistanceApplicationService: TravelDistanceApplicationService

    @BeforeEach
    fun setUp() {
        travelDistanceApplicationService = TravelDistanceApplicationService(distanceMatrixService, organizationRepository)
    }

    @Test
    fun searchReturnsValuesForAllPossibleTravelOptions() {
        `when`(distanceMatrixService.getTravelDistanceFor(safeAny(TravelModeDto::class.java), safeAny(GeoPoint::class.java), safeAny(GeoPoint::class.java))).thenReturn(DUMMY_TRAVEL_DISTANCE)
        val travelOptions = testSearch()
        Assertions.assertThat(travelOptions)
            .isNotNull
            .hasSize(TravelModeDto.values().size)
    }

    @Test
    fun onlyReturnDistancesForFoundRoutes() {
        `when`(distanceMatrixService.getTravelDistanceFor(safeEq(TravelModeDto.TRANSIT), safeAny(GeoPoint::class.java), safeAny(GeoPoint::class.java))).thenReturn(null)
        `when`(distanceMatrixService.getTravelDistanceFor(safeNot(safeEq(TravelModeDto.TRANSIT)), safeAny(GeoPoint::class.java), safeAny(GeoPoint::class.java))).thenReturn(DUMMY_TRAVEL_DISTANCE)
        val travelOptions = testSearch()
        Assertions.assertThat(travelOptions)
            .isNotNull
            .hasSize(TravelModeDto.values().size - 1)
        Assertions.assertThat(
            travelOptions.filter { (travelMode) -> travelMode === TravelModeDto.TRANSIT }
        )
            .isEmpty()
    }

    private fun testSearch(): List<TravelDistanceDto> {
        val address = Address.Builder()
            .setLocation(GeoPoint(49.0388109, 8.3433651))
            .build()
        val organizationId = OrganizationId()
        `when`(organizationRepository.findOne(organizationId.value)).thenReturn(Organization.Builder()
            .setDefaultAddress(address)
            .build())
        return travelDistanceApplicationService.requestTravelDistances(organizationId, GeoPoint(48.9808278, 8.4907565))
    }

    companion object {
        private val DUMMY_TRAVEL_DISTANCE = TravelDistanceDto(TravelModeDto.CYCLING, 31200L, 34200L)
    }
}

fun <T : Any> safeEq(value: T): T = eq(value) ?: value
fun <T : Any> safeNot(value: T): T = not(value) ?: value
fun safeAny(value: Class<GeoPoint>): GeoPoint = any(value) ?: GeoPoint(49.0, 8.0)
fun safeAny(value: Class<TravelModeDto>): TravelModeDto = any(value) ?: TravelModeDto.CYCLING